### ISSUE TITLE
Provide mechanism to trivially re-enable AF_VSOCK in the default seccomp profile

### ISSUE THREAD

**Issue by gaby** [moby/moby#44670]
Maintainer edit: Won't be accepted as is, but alternative mechanisms not involving --cap-add should be supported.

---

### Description

Recently several PR's were merged blocking calls to AF_VSOCK. These were backported without warning on a friday as a patch release. This caused a major headache where multiple client systems went down.

There's currently no documentation on how to enable these calls again, or a way give a container the ability to use them unless using priviledge. Forcing users to use a custom seccomp profile as the only way to enable this is not user friendly. 

Proposing the addition of a CAP for enabling AF_VSOCK on a container basis.

Example:

```bash
docker run -it --cap-add NET_VSOCK alpine:latest
```

Potential names:
- SYS_VSOCK
- NET_SOCKET
- NET_VSOCK
- VSOCK

Valid use case:
- AF_VSOCK enables networks using Libvirt and KubeVirt to create management tunnels into their Libvirt Domains. For example, we deploy core VM's with a systemd service allowing VSOCK to TCP connection into the local ssh daemon. The domains are running on a segregated network, so the only way to communicate with them is via VSOCK in combination with socat. On the host we dynamically enable these tunnels using socat running inside Docker containers.

Related PR's: #44562 #44563 #44564 

Cc. @thaJeztah @AkihiroSuda @GabrielNicolasAvellaneda

**Comment by AkihiroSuda** [comment]
If this is really expected to be a capability, this should be proposed to the kernel. But getting this accepted in the upstream might be difficult, as the kernel has been already consuming 41 of the total 64 bits.

Perhaps just adding a new seccomp option like `--security-opt seccomp=+vsock` might suffice?

**Comment by AkihiroSuda** [comment]
We can also consider defining the "virtual capability" like `--cap-add VCAP_NET_VSOCK` for controlling the seccomp profile, but that might be rather confusing.

**Comment by gaby** [comment]
@AkihiroSuda Trying to use `--security-opt` was my first though. After 10-20mins I realized it had to be backed in into the engine for it to work. The VCAP approach would make sense if there's other functions that could also benefit from having a Virtual Capability assigned to them.

**Comment by neersighted** [comment]
I strongly object to the concept of a virtual capability -- capabilities in the Linux kernel are well-known and well-defined, and inventing our own is overloading the meaning of the API field and going to create confusion in the ecosystem.

**Comment by neersighted** [comment]
Also, as there was no update here: This was discussed in the maintainers call. The consensus was no new flags will be added; instead the VSOCK change will be reverted in the 20.10 branch (keep in mind that it is **not namespaced** so your containers will be able to talk to the Hypervisor Guest Services socket for instance). On 23.0 and newer, it will be blocked by default and you will need to pass a custom seccomp filter if you want to re-allow it.

There are patches upstream that add namespacing to AF_VSOCK, but none have been accepted yet/look likely to land in the next several cycles, so this will likely remain the state of the art for some time.

**Comment by gaby** [comment]
@neersighted Totally agree, I think the best route is to add a seccomp filter instead of a Virtual Capability.

**Comment by neersighted** [comment]
Re-opening this; while it will not be implemented as-is, https://github.com/moby/moby/issues/47663 has me thinking of other ways to answer this ask.

Alternatively, https://github.com/moby/moby/issues/22109 is more involved but a more holistic way to so solve this in the long term.

**Comment by gaby** [comment]
@neersighted Agree, as of today with `v26.x` using a custom `seccomp` is the only way to re-enable `VSOCK` support.

**Comment by gaby** [comment]
@neersighted @thaJeztah Any updates on this? Everytime I have to update the `docker-ce` package I'm afraid of breaking the whole system. It also means I have to constantly check Docker release notes to see if the `policy.json` file changed.

**Comment by gaby** [comment]
Two years later and there's still no solution for this? Just because kubevirt asked for it to be removed, now everyone using this is stuck.

**Comment by gaby** [comment]
@neersighted @AkihiroSuda @thaJeztah Since `v28.0.2` or `v28.0.3`, i'm no longer able to use `VSOCK` at all even when passing a `profile.json`. Did something change?

Even when running as privilege, it won't work. I get the same error regardless of `profile.json` file when connecting to a VSOCK.
